### DISCUSSION TITLE
Impose a solution of block 1 onto block 2

### DISCUSSION THREAD

**Discussion by TLWise** [idaholab/moose#20884] [Q&A Modules: General] [answered]
I am solving a heat transfer problem in MOOSE. The problem domain solves for the temperature on two separate meshes(two blocks same file). I am imposing a constant temperature on a small circular domain. I am solving convective heat in the other mesh domain which is square and much larger than the circular domain. I would like to impose the solution of the constant temperature domain into the convective solution domain in order to determine the effect of the imposed constant solution on that of the convective solve. Is it possible to translate a nearest neighbor (meshes overlap) solution from one mesh domain into another without using MultiApp, since the meshes are contained in the same mesh file and are solved for in the same Kernel block?

**Comment by GiudGiud** [comment, marked as answer]
Hello

I dont know of a way to exactly do that currently.
If the 'nearest-neighbor' transformation (eg you are happy with that mapping of the square onto the circle) works for you, you could use a MultiApp to use that transfer. 

Else if there's a 1-1 relationship between nodes in the circle and square boundaries you could try using the Constraints system.

Finally you could write your own user object to do this. 

Guillaume

**Reply by TLWise** [reply]
I have implemented MultiAppNearestNodeTransfer where I have defined the physics and boundary conditions for each mesh (block) in separate solver files. Each solutions is a transient solution. The controlling app appears to progress as a time based solution. The multi app input file appears to restart for each data push and is not time varying. Is it possible preserve the transient solution of all files and observe the progression of the solution in the controlling app?

**Reply by GiudGiud** [reply]
So in a MultiApp set up?
Did you use a FullSolveMultiApp or a TransientMultiApp ? Because the former is resetting every time step currently. 
The latter is the one to use to get parallel time evolution of the two solves.

**Reply by TLWise** [reply]
I am using the TransientMultiApp, with the following format:
```
[MultiApps]
  [sub_app]
    type = TransientMultiApp
    positions = '0.1 0.1 0  0.4 0.4 0  0.7 0.7 0'
     input_files = '2D_Move_Fluid_Ball.i'
    execute_on = timestep_end
    output_in_position = true
  []
```

**Reply by GiudGiud** [reply]
the exodus output if turned on in the subapp should include the solution of the subapp at all times in the transient

**Reply by TLWise** [reply]
When I solve the problem,I am only seeing the main output and not the subapp. Exodus is true for both.

**Reply by GiudGiud** [reply]
what does `ls` give in the directory?

**Reply by TLWise** [reply]
peacock_run_exe_tmp_2D_Move_Ball_All_MultiApp.i
peacock_run_exe_tmp_2D_Move_Ball_All_MultiApp_out.e
peacock_run_exe_tmp_2D_Move_Ball_All_MultiApp_out_sub_app0.e
peacock_run_exe_tmp_2D_Move_Ball_All_MultiApp_out_sub_app1.e
peacock_run_exe_tmp_2D_Move_Ball_All_MultiApp_out_sub_app2.e
peacock_run_exe_tmp_2D_Move_Fluid_Ball.i
peacock_run_exe_tmp_2D_Move_Fluid_Ball_out.e
peacock_run_mesh_tmp_2D_Move_Ball_All_MultiApp.e

**Reply by TLWise** [reply]
I am seeing what I expect in the subapp_out, but that solution does not seem to translate back into the MultiApp_out.

**Reply by GiudGiud** [reply]
these ones:
```
peacock_run_exe_tmp_2D_Move_Ball_All_MultiApp_out_sub_app0.e
peacock_run_exe_tmp_2D_Move_Ball_All_MultiApp_out_sub_app1.e
peacock_run_exe_tmp_2D_Move_Ball_All_MultiApp_out_sub_app2.e
```
these are the outputs of the subapp.
What do you mean by translate back? As in have the solutions in the same file?

**Reply by TLWise** [reply]
Here are my files:
[2D_Move_Ball_All_MultiApp_i.txt](https://github.com/idaholab/moose/files/8588161/2D_Move_Ball_All_MultiApp_i.txt)
[2D_Move_Fluid_Ball_i.txt](https://github.com/idaholab/moose/files/8588162/2D_Move_Fluid_Ball_i.txt)
[FallingBall_e.txt](https://github.com/idaholab/moose/files/8588163/FallingBall_e.txt)

**Reply by GiudGiud** [reply]
oh you mean the transfer is failing in the direction subapp -> main app?

**Reply by GiudGiud** [reply]
I see that this transfer is towards `T`, which is a main variable. So whatever is transferred gets overwritten by the solve in the main application. Like what is transferred becomes an initial guess, then when the solve happens that is replaced by the solution. 

Is this the transfer you are talking about?

**Reply by TLWise** [reply]
Yes. How do I get a combination of the sub_app solution to serve as an initial condition of the main_app next transient step solution?

**Reply by GiudGiud** [reply]
What you are doing is one way to do this. 
The subapp is executed on timestep_end. so at the end of time step n-1, the solution of the subapp is sent to the main app.
When the main app solves time step n, it starts from the temperature field previously transfered. 

You can have a look at the log to see when things are being transfered.
You can also set `solve=false` in [Problem] in the main app, which will cancel the solve. Then you should only see what is being received from the subapp

**Reply by TLWise** [reply]
Is it possible to couple the value of an aux variable with a variable?

**Reply by GiudGiud** [reply]
what do you mean by couple here? Could you write down a numerical scheme so I can understand?
They can be coupled in very many ways in MOOSE

**Reply by GiudGiud** [reply]
@TLWise any luck with this?

**Reply by TLWise** [reply]
Not as of yet. I am doing additional reading of the literature to gain an understanding of how solutions with weak couplings between domains has been implemented.

**Reply by GiudGiud** [reply]
the functor system could be used to move data from one point to another point with arbitrary relations btw. 
But I doubt the rest of your simulation will be compatible. What modules are you using?

**Reply by TLWise** [reply]
I'm using Combined to solve heat transfer, fluid flow, and newton motion.

**Reply by GiudGiud** [reply]
FE fluid flow or FV?

**Reply by TLWise** [reply]
FE. I'm trying to remain a MOOSE purists.

**Reply by GiudGiud** [reply]
ok then functors are not an option. 

A MultiApp setup seems the more natural here.
Second a custom auxkernel / UO could do this, but it'd be a little hacky. You d have to write your mapping explicitly, use a point locator to locate the source data, re-init there to compute the Qp values if they are needed, then transfer the data.

**Reply by GiudGiud** [reply]
I guess we drifted here. 

You also had an issue with the transfer from the subapp to the main app.
What is the extent (XY coordinates) of the mesh in both apps?
Do you need to use a position argument in the MultiApp to make them match?

Guillaume

**Reply by TLWise** [reply]
I am using the same mesh file for each solver app. There are two mesh domains (2D x,y) contained within one exodus file, one mesh (a circle) initially defined within the boundary of the other mesh (a tall rectangle). Slow motion of the circle with motion in the y-direction stays within the boundary of the rectangle.

One app solves heat transfer on the rectangular mesh and is constrained against motion. The other app solves heat transfer (constant temperature) and newton motion due to gravity on the circular mesh. My intent is to show the effects of the circle held at constant higher temperatures, boundary effects on the solutions on the temperature of the tall rectangular domain. I want to translate the circles temperature in the domain of the rectangle as it moves under the influence of gravity.

**Reply by GiudGiud** [reply]
so you are using BlockDeletionGenerators to tailor each solver app's mesh ?

are you looking for some sort custom transfer to do this?
I think you can try nearest node if you translate one of the domains but otherwise not sure what else without making a custom object

**Reply by TLWise** [reply]
I used NullKernel in the apps that were not required to solve the physics in the extra mesh domain. I will apply BlockDeletionGenerator to see what that results in. Thank you.

**Reply by GiudGiud** [reply]
NullKernel would work but it s more costly, both in memory and in terms of making the matrix bigger because of unused Dofs
